### PR TITLE
std dirs cdhook and tests - fix to track CD

### LINKED ISSUE #9229: cd does not update std dirs
### Describe the bug

If you `cd` after entering a dir with std dirs it foes not update. Same with shells

### How to reproduce

1. Open a nushell instance
2. Run `shells`
3. Cd into another dir
4. Run shells again

### Expected behavior

The last shells call should show the current dir, but it shows the last

### Screenshots

_No response_

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.80.1                                   |
| branch             | main                                     |
| commit_hash        | 6a0c88d516faf9fd1165fddbd7ef69ba47f97cc6 |
| build_os           | windows-x86_64                           |
| build_target       | x86_64-pc-windows-msvc                   |
| rust_version       | rustc 1.67.1 (d5a82bbd2 2023-02-07)      |
| rust_channel       | 1.67.1-x86_64-pc-windows-msvc            |
| cargo_version      | cargo 1.67.1 (8ecd4f20a 2023-01-10)      |
| build_time         | 2023-05-17 18:54:25 -05:00               |
| build_rust_channel | release                                  |
| features           | default, zip                             |
| installed_plugins  | periodic-table                           |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bobhy**
# Description

[rewriting history again, this post describes the final, final result]

Provides a PWD change handler to keep `std dirs` (and aliases `shell` and `enter`) in sync with the 
current working directory (`$env.PWD`) when  `cd` command changes it.

Fixes #9229.

# User-Facing Changes

You must manually configure the event handler for PWD environment variable changes.  In your `config.nu`:
```
let-env config {
    . . .
    hooks: {
        env_changed: {
            # before: PWD: [{|before, after| null}]
            PWD: [{|before, after| std dirs cdhook $before $after}]
        }
    }
    . . .
}

```

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by bobhy** [comment]
Tried to add this hook to the default config, but the unit tests on that file  failed  for deeply mysterious reasons, I'll try to narrow down the repro scenario and post an issue, in the meantime, I left  some info on discord.
So *all* users will have to edit config file by hand in the meantime.

I looked  implementing my own wrapper for `cd` command instead of the hook approach. It actually seemed possible, but all users would be forced to use it, and any glitch in behavior of `cd` would be highly visible.  I didn't want to risk it.

This way, if the hook malfunctions, the worst outcome is the current behavior (confusing display of directory ring), but at least you would be able to change directories!

**Review by amtoine — CHANGES_REQUESTED** [review]
as @kubouch suggested, i think this can be achieve with modifying the `export-env` block of `std dirs`.

i'm not liking the fact that these changes require the user to do anything, this should work out of the box for them :yum: 

i'm not happy with the `default` part of it, but the following appears to work as the OP expected :ok_hand: 
```bash
export-env {
    let-env DIRS_POSITION = 0
    let-env DIRS_LIST = [($env.PWD | path expand)]

    $env.config = ($env.config? | default {})
    $env.config.hooks = ($env.config.hooks? | default {})
    $env.config.hooks.env_change = ($env.config.hooks.env_change? | default {})
    $env.config.hooks.env_change.PWD = ($env.config.hooks.env_change.PWD? | default [])

    $env.config.hooks.env_change.PWD = ($env.config.hooks.env_change.PWD | append [
        {|before, after|
            if not ($after | is-empty) {
                let-env DIRS_LIST = ($env.DIRS_LIST | update $env.DIRS_POSITION $after)
            }
        }
    ])
}
```

EDIT: this works with `cargo run -- -n`

**Comment by kubouch** [comment]
Wow, the `default` chain is really ugly, we need to figure out something to make inserting into nested records easier.

One problem with this is that if you run the `export-env` repeatedly, it will keep inserting the PWD hook to the list. Maybe we could check with `view source` whether the PWD hook list already contains the block. Something like:
```
let closure = {|before, after| ... }

let already_has_hook = ($env.config.hooks.env_change.PWD
	| any {|hook|
		(view source $hook) == (view source $closure)
	})
	
if not $already_has_hook { ... add hook }
```

**Comment by kubouch** [comment]
And yes 

> users just need to make sure they do not overwrite the $env.config.hooks block in their config.

is a good point. We should change the default config such that it appends to the existing hook list instead of overwriting, and put it to the release notes as a breaking change warning.

**Comment by amtoine** [comment]
agree with all your points @kubouch :+1: 
in the end, i do not think the current PR is the direction to take :thinking:

**Comment by amtoine** [comment]
i've added the `breaking change` label, just to avoid forgetting about the potential breaking part of this :relieved:

**Comment by kubouch** [comment]
I'll try later today fixing the `alias` predecl lookup so that the "wrapper" solution could be used. It's not a guarantee, though, there might be some hidden side effects.

**Comment by fdncred** [comment]
it would also be good if we could eventually change the `default` command to create every level if it doesn't exist so you only have to do this
```
$env.config.hooks.env_change.PWD = ($env.config.hooks.env_change.PWD? | default [])
```
and not create a default for $env.config, $env.config.hooks, $env.config.hooks.env_change, etc.

one last question, does `dirs cdhook` do the thing that was missing, when enabled? i.e. does it update the list with the current folder as active = true and set all other items in the shells list as active = false?

**Comment by amtoine** [comment]
agree with the `default` part of your comment, definitely :ok_hand: 

> does `dirs cdhook` do the thing that was missing, when enabled?
>
i haven't tested :thinking: 

the pure `export-env` change does work however :+1:

**Comment by bobhy** [comment]
@amtoine, specifically what approach do you think this PR should be taking?
I don't think it will be practical at the end of the day to provision the hook automatically, but I can go back to
seeing whether I can alias `cd` itself (while preserving all the special features of builtin cd).  I think I can see a way there...

Re using the export env block, two issues:
1. Make the update idempotent, because the block gets executed every time you `use` the module.  A small tweak to the snippet from @kubouch would accomplish that.
2. But we're hiding what should be optional configuration from the user.  S/he has no ability to modify the automatic registration, s/he doesn't see it in `config.nu`.   Maybe we need to extend the `config` command to actually modify the configuration (at runtime)?

**Comment by bobhy** [comment]
This looks pretty solid.  I dropped the whole hook idea and made a wrapper for the builtin CD command which is globally aliased to `cd` in the prelude.  So I'm making a very core function of the interpreter (cd) dependent on this stdlib extension, which is a fairly bold step.
I was careful to wrap cd rather than reimplement it, so you still get cd tricks like abbreviated paths.
I was concerned about the impact of loading the module std dirs multiple times perhaps somehow invalidating the cd alias, but havn't seen any evidence of it, and there weren't any problems in the test runs, fwiw.
The only thing I can think we might need is another global alias such as `core_cd` to access the builtin cd when all else fails.  I'd hate to be stranded in an interpreter that can't change directories.  Not sure where to put that, though.

**Comment by amtoine** [comment]
@bobhy 
does the user have anything to do now to have both the fix and the original `cd` working as they should?

**Comment by kubouch** [comment]
Check this https://github.com/nushell/nushell/pull/9244 . Seems you've managed to find a workaround though.

**Comment by fdncred** [comment]
nice, are we ready to land this?

**Comment by amtoine** [comment]
just a sec :+1:

**Comment by amtoine** [comment]
> nice, are we ready to land this?
>
if we're ok with the breaking change on the builtin `cd` command, yup, that looks ready :+1:

**Comment by fdncred** [comment]
what's the breaking change on the builtin cd command?

**Comment by amtoine** [comment]
> what's the breaking change on the builtin cd command?
>
ooh :eyes: 

actually Darren, you might want to build the branch quickly and try out `help cd` and `help builtin cd` :grimacing: 


TL;DR: built-in `cd` is kinda gone with this PR, as far as i understand

**Comment by fdncred** [comment]
I guess what you mean is this
```
export alias "builtin cd" = cd
export alias cd = dirs cd
```

I'm trying to remember how it used to work since cd updated the active = true. Did we have code in cd to do that?

**Comment by amtoine** [comment]
yeah, but i also mean the the built-in `cd` command can not be accessed anymore.
it's through an alias so we lose its `help` page and options

**Comment by fdncred** [comment]
i'm not sure about this then. I achieved this screenshot in an unnatural way. I did `hide cd` and then cd'd somewhere, which uses the external cd command. So, if I `hide cd` that removes the script that updates this. Clearly I'm breaking this on purpose. 
(i'm not in the nushell folder, i'm in the home folder)
image

I'm wondering what other alternatives we have. I really don't like losing the `help cd`.

Do we need to check to make sure a folder isn't already in the list?
(nu-cli is in the list twice, is that right?)
image

Another thought I've had in the past is that it would be nice to have some way to indicate that I want to run a builtin command. Maybe a sigil or something like we have ^ for externals.

I made this draft until we work through the remaining issues. More input from the team would be helpful.

**Comment by amtoine** [comment]
i proposed the *hook* thing above which worked but just had an ugly `default` pattern :thinking: 

> **Note**
> see [here](https://github.com/nushell/nushell/pull/9234#pullrequestreview-1434071794)

but it looked to be working at least, without overwriting the builtin `cd`

**Comment by fdncred** [comment]
if the hook fixes this issue of overriding a builtin command, then i'd vote for that. it probably also fixes my first screenshot, but maybe not the second one.

it seems like the shells/enter command from 0.79.0 allowed us to duplicate entries too, which seems like a bug to me that maybe we could fix in this script.
image

**Comment by bobhy** [comment]
We're dithering here!  Let's get clear on the problems we actually want to fix and fill this or kill it.
1. Going back to using the hook means users have to update config to enable it, (or we have to do something magical to update their configuration for them).
2. Fixing `cd help` seems doable using the current approach.  Reverting the extra alias `dirs cd` probably fixes the `help builtin cd` issue.  Might only need to change the *ordering* of aliases, given that recent predecl change.
3. Rethinking whether shells or its replacement ought to dedup the list seems out of scope of this PR, I think.
4. Finally, we could abandon `dirs cd` and the need to wrap cd entirely by modifying the builtin `cd` to update the environment variable `DIRS_LIST` (if it finds it).  
5. We could even implement the rest of CD functionality in the script (OLDPWD, abbreviations), it's not  that complex a command.  I wouldn't recommend this approach, I think CD functionality should be a builtin.

I like #2, and will spend an hour trying to make it work.  If you all think that even a perfect wrapping of CD would not be an acceptable approach, please let me know...

**Comment by bobhy** [comment]
btw, `help commands builtin cd` is doing the right thing before I make any further changes.  Why is builtin help fixating on the (unhelpful) alias definition?
```
help commands builtin cd
Change directory.

Search terms: change, directory, dir, folder, switch

Usage:
  > builtin cd (path) 

Flags:
  -h, --help - Display the help message for this command

Signatures:
  <nothing> | builtin cd <string?> -> <nothing>
  <string> | builtin cd <string?> -> <nothing>

Parameters:
  path <directory>: the path to change to (optional)

Examples:
  Change to your home directory
  > cd ~

  Change to a directory via abbreviations
  > cd d/s/9

  Change to the previous working directory ($OLDPWD)
  > cd -

〉help builtin cd
Change directory.

Alias: builtin cd

Expansion:
  cd
```

**Comment by fdncred** [comment]
> Going back to using the hook means users have to update config to enable it,

Can't we just add the hook to the script so that it appends the hook and not overwrite people's hooks? If we could do that, I think the hook would work better than aliasing a builtin command.

**Comment by amtoine** [comment]
yes, this is precisely what i'm trying to show in the `export-env` block i shared above :+1:

**Comment by fdncred** [comment]
> yes, this is precisely what i'm trying to show in the `export-env` block i shared above 👍

That's what I thought so I don't understand @bobhy's objection.

**Comment by bobhy** [comment]
Sure, it's just software.
I think the last suggestion on that thread was to do the config update in the `export def-env` block, to append to whatever's there and to make sure there's only a single entry, no matter how many times the module was 'use'd?  (which does happen in tests, even if not in real life).

**Comment by bobhy** [comment]
Reverted to hook impl, but doesn't work at first prompt because the hook hasn't been defined.
Also test test_dirs_cd shows more detail on why it's failing -- somehow $env.config gets overwritten after being set up in export def-env.  Please advise!

**Comment by bobhy** [comment]
The idea of automatically updating the configuration from the loading of the stdlib module seems to run afoul of 2 things:
1. I think that the user's config file is sourced *after* the stdlib is loaded and prelude is run.  For this PR, that means that nu startup runs `use std dirs`, which runs my `def-env` which sets up the PWD hook, but later, nu startup sources `config.nu` and that overwrites whatever changes I had made to `$env.config`.  Does that sound likely to you all?  Diving into the code now to see if I can confirm this.
2. For sure, all of `$env.config` is not defined when `testing  run-tests` invokes stdlib tests.  It does this by `nu -c <wrapper-to-test.nu>` and that definitely starts nu without `$env.config`, to wit:

```
〉nu -c 'print $env.config.history.max_size'
Error: nu::shell::column_not_found

  × Cannot find column
   ╭─[source:1:1]
 1 │ print $env.config.history.max_size
   ·       ──┬─ ───┬──
   ·         │     ╰── cannot find column 'config'
   ·         ╰── value originates here
   ╰────

```
(which is exactly the issue causing my test case to fail).  
It is seeing #2 happen that makes me think #1 is likely.

In the short run, I think this means we must require users to make a manual edit to their own `config.nu` to define the hook.  If you really hate that, the next best thing I can think of is to modify the builtin CD to be aware of `dirs` and for it to directly update environment variable $env.DIRS_LIST.  That's a bit ad-hoc, but technically easy.

In the longer run, I think we can rationalize the sequence of operations at startup to make the current approach work as we expected.  Perhaps the path of least resistance is to defer initialization of the standard library until after the user's `config.nu` has run?  

Plus, what to do about the way `run-test` invokes test cases?  As it's currently implemented, stdlib will *never* be able to test *any feature*  involving configuration changes.

**Comment by bobhy** [comment]
Yah, #1 above is confirmed so far as timing.  The sequence is driven from `main.rs`.  it acquires the terminal, then (line 156) loads standard library (which executes our `def-env` which updates `$env.config`), then later (line 258) decides whether it's a REPL invocation.  Only if it's a REPL invocation will the user configuration files be loaded (`run.rs, ln 208).  

The merge of user config into engine_state via `.into_config()` is coarse grained.  If user config has a `config.hook` field (which it does, by default), the whole hook is copied into engine_state.config.hook; no attempt is made to see whether engine_state.config.hook already had something, or to merge, e.g individual array elements from the PWD hook list to the other. (assuming that the `def-env` changes got  copied from $env.config to engine_state.config in the first place, which I haven't confirmed)

So the automatic hook config is going to need architectural fixes to be viable.  In the meantime, I'm going to post a small change to make it easy for user to manually configure the hook and let you all decide how to resolve this issue.

**Comment by bobhy** [comment]
Release notes at top updated, tests pass.  Once configured, this works fine.  I'm sure we'll have an opportunity to improve on it in the future!
Don't let the perfect be the enemy of the good!

**Comment by kubouch** [comment]
@bobhy The current solution requiring user to manually set up a hook has the problem that buggy shells commands are loaded by default, and user is supposed to fix it by adding a hook. This is a situation we'd like to avoid, we chatted about it and reached the following consensus:
* Implement the shells update by wrapping the `cd` builtin command (this solution seems simpler than trying to modify hooks).
* Do not load the shells commands by default (i.e., remove them from the prelude)

Since the shells commands do not seem to be used as often, not having them by default should not be such a problem and `cd` would not be overriden by default. Overriding `cd` would be an opt-in action when importing `dirs` and can be explained in the documentation of the module.

Sorry for going back and forth with this.

A side note: Seems like the main concern about overriding the builtin `cd` command was insufficient help message: missing signatures, examples, search terms (and these should be colorful). At some point, we might think about creating a doc comment format (like Rust's `//!`) that would allow us to specify these missing features and reach feature parity with builtin commands' help messages.

**Comment by bobhy** [comment]
I can go back to the override cd approach and can at least make `help
commands cd` do the right thing, even if `help cd` itself does not.

We've been down 2 roads on this fix and confirmed there is not a perfect
solution down either one.
Here's a third option that might be better: have the built in `cd` check
for `DIRS_LIST` and update it if it finds it.  CD is already updating
OLDPWD as well as PWD itself, it knows about environment variables.
This approach does create an ugly cross dependency, but that's not a
functional problem.
Any interest?

**Comment by bobhy** [comment]
Went back to the original problem report and focused on solving the problem as stated.  That led to a much cleaner solution: let CD do its thing and have `std dirs` show (and use) the real `$env.PWD` rather than its own stale copy in the active slot of `$env.DIRS_LIST`. No hooks, no wrapping or overriding `CD`, just clean separation of interests!

See #9267

**Comment by hyiltiz** [comment]
Is it similar to #9234? I tried to work out the rebase in  https://github.com/hyiltiz/nushell/tree/fix_9229 but it looks like that PR was stale?

**Comment by amtoine** [comment]
what do you mean @hyiltiz?
i remember this PR went *out of control* and was thus close in favor of #9267 :thinking:

**Comment by hyiltiz** [comment]
Ignore me then; I tried to rebase but got confused by the many re-PRs of this feature/fix, and wasn't sure where was the latest state of work. Thanks for the pointer!
